consensus/ibft/fork/manager.go

```diff
@@ -0,0 +1,434 @@
+	secretManager secrets.SecretsManager,
+	filePath string,
+	epochSize uint64,
+	ibftConfig map[string]interface{},
+) (ForkManager, error) {
+	forks, err := GetIBFTForks(ibftConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	fm := &forkManagerImpl{
+		logger:         logger.Named(loggerName),
+		blockchain:     blockchain,
+		executor:       executor,
+		secretsManager: secretManager,
+		filePath:       filePath,
+		epochSize:      epochSize,
+		forks:          forks,
+		signers:        make(map[validators.ValidatorType]signer.Signer),
+		validatorSets:  make(map[store.SourceType]store.ValidatorStore),
+	}
+
+	// Need initialization of signers in the constructor
+	// because hash calculation is called from blockchain initialization
+	if err := fm.initializeSigners(); err != nil {
+		return nil, err
+	}
+
+	return fm, nil
+}
+
+// Initialize initializes ForkManager on initialization phase
+func (m *forkManagerImpl) Initialize() error {
+	if err := m.initializeValidatorStores(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetSigner returns a proper signer at specified height
+func (m *forkManagerImpl) GetSigner(height uint64) (signer.Signer, error) {
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	signer, ok := m.signers[fork.ValidatorType]
+	if !ok {
+		return nil, ErrSignerNotFound
+	}
+
+	return signer, nil
+}
+
+// GetValidatorStore returns a proper validator set at specified height
+func (m *forkManagerImpl) GetValidatorStore(height uint64) (store.ValidatorStore, error) {
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	set, ok := m.validatorSets[ibftTypesToSourceType[fork.Type]]
+	if !ok {
+		return nil, ErrValidatorStoreNotFound
+	}
+
+	return set, nil
+}
+
+// GetValidators returns validators at specified height
+func (m *forkManagerImpl) GetValidators(height uint64) (validators.Validators, error) {
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	set, err := m.GetValidatorStore(height)
+	if err != nil {
+		return nil, err
+	}
+
+	return set.GetValidators(
+		calculateFetchingValidatorsHeight(height, m.epochSize, fork),
+	)
+}
+
+// GetHooks returns a hooks at specified height
+func (m *forkManagerImpl) GetHooks(height uint64) (hook.Hooks, error) {
+	hooks := &hook.HookManager{}
+
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	var err error
+
+	switch fork.Type {
+	case PoA:
+		err = m.registerPoAHooks(hooks, height)
+	case PoS:
+		registerPoSHook(hooks, m.epochSize)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := m.registerPoAPrepareHooks(hooks, height); err != nil {
+		return nil, err
+	}
+
+	m.registerPoSPrepareHooks(hooks, height)
+
+	return hooks, nil
+}
+
+// Close calls termination process of submodules
+func (m *forkManagerImpl) Close() error {
+	if err := m.closeSnapshotValidatorStore(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
```

Comment:
Each `Store` implementation should have the `Close` method implemented. In the case of a contract store, it would be a no-op.

Reply:
As I explained above, I want to keep SnapshotStore pure data structure. And I've thought that this module's responsibilities are to absorb the differences and provide module with common interface to IBFT

Reply:
Alright, but then you have to constantly deal with type assertions ([like this](https://github.com/0xPolygon/polygon-edge/blob/a5ddd27d319ce4a0a96cb496c4db33469413c25f/consensus/ibft/fork/manager.go#L282)) whenever you care about snapshots in particular. If side effects such as i/o read-write fall onto the `forkManager` then why not just  create separate fields and manage each store separately like so:

```
	validatorSets map[store.SourceType]store.ValidatorStore	// not this
	
	snapshotStore snapshot.SnapshotValidatorStore
	conractStore contract.ContractValidatorStore
```

Also, another thing worth noting is that the snippet pasted above was copied from the actual codebase I just edited (no IDE errors). Meaning that, if I'm able to see the concrete implementations of the `store.ValidatorStore` type, then why bother with the interface in the first place since I'm not hiding any implementation ?

The moment you need to know the concrete implementation is when the interface stops making sense and code is losing on uniformity. I'm not saying the design is not functional, but you get why it is hard on the eyes and harder to reason about.

Feel free to close this comment if you feel the suggestion would add complexity:

**[Suggestion]** Uniform interface that both validator store types (`Snapshot`, `Contract`) will implement. Consequentially, these 2 implementations would reside in their own package offering their `Store` implementations via `NewXXX() store.ValidatorStore` constructor.

Reply:
> Also, another thing worth noting is that the snippet pasted above was copied from the actual codebase I just edited (no IDE errors). Meaning that, if I'm able to see the concrete implementations of the store.ValidatorStore type, then why bother with the interface in the first place since no implementation is actually being hidden?
The moment you need to know the concrete implementation is when the interface stops making sense and code is losing on uniformity. I'm not saying the design is not functional, but you get why it is hard on the eyes and harder to reason about.

There is one thing you're misunderstanding. ForkManager calls common methods which are defined in interface in both store. Only the differences FookManager needs to care about is initialization and termination. It was possible to implement Wrapper class that has Close() method but I had no big reason to implement that.

I created Wrapper class under fork package 41b2f5f9c008618b60ec0d8ced5288d330c9d423

Reply:
@dbrajovic Is that what you expects?